patches/server/0887-Better-Stats-API.patch

```diff
@@ -0,0 +1,471 @@
++        } else {
++            throw new IllegalArgumentException("Did not recognize " + Registry.STAT_TYPE.getKey(type) + " as a statistic type");
++        }
++        return Objects.requireNonNull(nmsStat, "Couldn't convert " + paperStat + " into a nms stat");
++    }
++}
+diff --git a/src/main/java/net/minecraft/stats/ServerStatsCounter.java b/src/main/java/net/minecraft/stats/ServerStatsCounter.java
+index c59f98ca3adfdd90cdc5999aeadbb0834efedc0f..a188b93e8b413571ce787c45842f7b9760b0f996 100644
+--- a/src/main/java/net/minecraft/stats/ServerStatsCounter.java
++++ b/src/main/java/net/minecraft/stats/ServerStatsCounter.java
+@@ -247,6 +247,20 @@ public class ServerStatsCounter extends StatsCounter {
+ 
+             object2intmap.put(statistic, this.getValue(statistic));
+         }
++        // Paper start
++        if (io.papermc.paper.event.player.PlayerRequestStatisticsEvent.getHandlerList().getRegisteredListeners().length > 0) {
++            io.papermc.paper.event.player.PlayerRequestStatisticsEvent statEvent = new io.papermc.paper.event.player.PlayerRequestStatisticsEvent(
++                player.getBukkitEntity(),
++                object2intmap.object2IntEntrySet()
++                    .stream()
++                    .collect(Object2IntOpenHashMap::new, (map, entry) -> map.put(io.papermc.paper.statistics.PaperStatistics.getPaperStatistic(entry.getKey()), entry.getIntValue()), Object2IntOpenHashMap::putAll)
++            );
++            if (!statEvent.callEvent()) {
++                return;
++            }
++            object2intmap = statEvent.getStatisticMap().object2IntEntrySet().stream().collect(Object2IntOpenHashMap::new, (map, entry) -> map.put(io.papermc.paper.statistics.PaperStatistics.getNMSStatistic(entry.getKey()), entry.getIntValue()), Object2IntOpenHashMap::putAll);
++        }
++        // Paper end
+ 
+         player.connection.send(new ClientboundAwardStatsPacket(object2intmap));
+     }
+diff --git a/src/main/java/org/bukkit/craftbukkit/CraftOfflinePlayer.java b/src/main/java/org/bukkit/craftbukkit/CraftOfflinePlayer.java
+index 1f2bc88d4570c6ef00e67a772b745e0b0c98e051..b2912ee3c8a8201146cf8d4bdca7e721b2c0f067 100644
+--- a/src/main/java/org/bukkit/craftbukkit/CraftOfflinePlayer.java
++++ b/src/main/java/org/bukkit/craftbukkit/CraftOfflinePlayer.java
+@@ -337,6 +337,69 @@ public class CraftOfflinePlayer implements OfflinePlayer, ConfigurationSerializa
+         return this.server.getHandle().getPlayerStats(this.getUniqueId(), this.getName());
+     }
+ 
++    // Paper start
++    @Override
++    public void incrementStatistic(io.papermc.paper.statistics.Statistic<?> statistic) {
++        this.incrementStatistic(statistic, 1);
++    }
++
++    @Override
++    public void decrementStatistic(io.papermc.paper.statistics.Statistic<?> statistic) {
++        this.decrementStatistic(statistic, 1);
++    }
```

Comment:
I don't see the point of that, sound just API bloat like bukkit does in its stat API

Reply:
I've gotten conflicting comments on using default methods. I know some people think all methods should be implemented in the impl class, rather than using default methods on the interface because it can make it confusing when only looking at an API class. You see a default method, but you don't know that it is also overridden some places.

Reply:
I'm not sure to understand the argument here, but the benefit of the default method here is to separate and isolate the api from the core implementation, and then if you search the method in the implementation we will not find the default parameter but only the "real" method. Also this avoid you to create redundant default value into the offline/online implementation (two times)